File: subscription-functions-copy/bulk-unsubscribe.py
#!/usr/bin/env python3
"""
Bulk Unsubscribe Script

Reads a CSV file with emails and unsubscribes them from a specified brand.

CSV Format:
- Must have an 'email' column
- Optional 'brand' column (if not provided, uses the --brand argument)

Usage:
    python3 bulk-unsubscribe.py emails.csv --brand heebnewsletters
    python3 bulk-unsubscribe.py emails.csv --brand heebnewsletters --batch-size 10
"""

import csv
import requests
import time
import argparse
import sys

# Default configuration
BASE_URL = "https://subscription-functions.vercel.app"
UNSUBSCRIBE_URL = f"{BASE_URL}/api/bulk-unsubscribe"

def unsubscribe_batch(emails, brand, dry_run=False):
    """Unsubscribe a batch of emails from a brand"""
    if dry_run:
        print(f"  [DRY RUN] Would unsubscribe {len(emails)} emails from {brand}")
        return {"success": True, "dry_run": True, "count": len(emails)}
    
    try:
        # Call the bulk unsubscribe endpoint
        payload = {
            "emails": emails,
            "brand": brand
        }
        
        response = requests.post(
            UNSUBSCRIBE_URL,
            json=payload,
            headers={"Content-Type": "application/json"},
            timeout=60
        )
        
        if response.status_code == 200:
            result = response.json()
            return {
                "success": True,
                "status_code": response.status_code,
                "result": result
            }
        else:
            return {
                "success": False,
                "status_code": response.status_code,
                "error": response.text[:200]
            }
    except requests.exceptions.Timeout:
        return {"success": False, "error": "Request timeout"}
    except Exception as e:
        return {"success": False, "error": str(e)}


def main():
    parser = argparse.ArgumentParser(
        description="Bulk unsubscribe emails from a brand",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Unsubscribe all emails in CSV from heebnewsletters
  python3 bulk-unsubscribe.py emails.csv --brand heebnewsletters
  
  # Process in batches of 10 with 1 second delay
  python3 bulk-unsubscribe.py emails.csv --brand heebnewsletters --batch-size 10 --delay 1
  
  # Dry run to see what would be unsubscribed
  python3 bulk-unsubscribe.py emails.csv --brand heebnewsletters --dry-run
        """
    )
    
    parser.add_argument("csv_file", help="Path to CSV file with emails")
    parser.add_argument("--brand", required=True, help="Brand ID to unsubscribe from")
    parser.add_argument("--batch-size", type=int, default=1, help="Number of emails to process per batch (default: 1)")
    parser.add_argument("--delay", type=float, default=0.5, help="Delay in seconds between batches (default: 0.5)")
    parser.add_argument("--dry-run", action="store_true", help="Preview what would be unsubscribed without actually doing it")
    parser.add_argument("--email-column", default="email", help="Name of the email column in CSV (default: 'email')")
    parser.add_argument("--brand-column", help="Name of the brand column in CSV (if different per email)")
    
    args = parser.parse_args()
    
    # Read CSV file
    print(f"📖 Reading CSV file: {args.csv_file}")
    emails_to_unsubscribe = []
    
    try:
        with open(args.csv_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            
            # Check if email column exists
            if args.email_column not in reader.fieldnames:
                print(f"❌ Error: CSV file must have a column named '{args.email_column}'")
                print(f"   Available columns: {', '.join(reader.fieldnames)}")
                sys.exit(1)
            
            for row_num, row in enumerate(reader, start=2):  # Start at 2 (row 1 is header)
                email = row.get(args.email_column, '').strip()
                if not email:
                    continue
                
                # Use brand from CSV if brand_column is specified, otherwise use --brand argument
                brand = row.get(args.brand_column, '').strip() if args.brand_column else args.brand
                if not brand:
                    brand = args.brand
                
                emails_to_unsubscribe.append({
                    "email": email,
                    "brand": brand,
                    "row": row_num
                })
    except FileNotFoundError:
        print(f"❌ Error: File not found: {args.csv_file}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Error reading CSV file: {str(e)}")
        sys.exit(1)
    
    total = len(emails_to_unsubscribe)
    if total == 0:
        print("❌ No emails found in CSV file")
        sys.exit(1)
    
    batches = (total + args.batch_size - 1) // args.batch_size
    
    print(f"📊 Found {total} emails to unsubscribe")
    if args.brand_column:
        print(f"📦 Brands will be read from CSV column: {args.brand_column}")
    else:
        print(f"📦 Brand: {args.brand}")
    print(f"⚙️  Batch size: {args.batch_size}, Delay: {args.delay}s")
    if args.dry_run:
        print("🔍 DRY RUN MODE - No actual unsubscribes will be performed")
    print()
    print("Starting processing...")
    print("=" * 60)
    
    results = {
        "total": total,
        "successful": 0,
        "failed": 0,
        "errors": []
    }
    
    start_time = time.time()
    
    # Group by brand for efficiency (if not using brand column)
    if not args.brand_column:
        # All same brand - process in batches
        all_emails = [item["email"] for item in emails_to_unsubscribe]
        all_brand = args.brand
        
        for i in range(0, total, args.batch_size):
            batch_emails = all_emails[i:i + args.batch_size]
            batch_num = (i // args.batch_size) + 1
            
            print(f"\n📦 Batch {batch_num}/{batches} ({len(batch_emails)} emails)")
            
            result = unsubscribe_batch(batch_emails, all_brand, dry_run=args.dry_run)
            
            if result.get("success"):
                if args.dry_run:
                    results["successful"] += len(batch_emails)
                else:
                    # Parse the API response
                    api_result = result.get("result", {})
                    summary = api_result.get("summary", {})
                    results["successful"] += summary.get("successful", 0)
                    results["failed"] += summary.get("failed", 0)
                    if "errors" in summary:
                        results["errors"].extend(summary["errors"])
                    print(f"      ✅ Batch processed: {summary.get('successful', 0)} successful, {summary.get('failed', 0)} failed")
            else:
                results["failed"] += len(batch_emails)
                error_msg = result.get("error", f"Status code: {result.get('status_code', 'unknown')}")
                for email in batch_emails:
                    results["errors"].append({
                        "email": email,
                        "error": error_msg
                    })
                print(f"      ❌ Batch failed: {error_msg}")
            
            # Delay between batches (except for the last batch)
            if i + args.batch_size < total and not args.dry_run:
                time.sleep(args.delay)
    else:
        # Different brands per email - process individually
        for i in range(0, total, args.batch_size):
            batch = emails_to_unsubscribe[i:i + args.batch_size]
            batch_num = (i // args.batch_size) + 1
            
            print(f"\n📦 Batch {batch_num}/{batches} ({len(batch)} emails)")
            
            # Group by brand within batch
            by_brand = {}
            for item in batch:
                brand = item["brand"]
                if brand not in by_brand:
                    by_brand[brand] = []
                by_brand[brand].append(item["email"])
            
            # Process each brand group
            for brand, emails in by_brand.items():
                print(f"  Processing {len(emails)} emails for brand: {brand}")
                result = unsubscribe_batch(emails, brand, dry_run=args.dry_run)
                
                if result.get("success"):
                    if args.dry_run:
                        results["successful"] += len(emails)
                    else:
                        api_result = result.get("result", {})
                        summary = api_result.get("summary", {})
                        results["successful"] += summary.get("successful", 0)
                        results["failed"] += summary.get("failed", 0)
                        if "errors" in summary:
                            results["errors"].extend(summary["errors"])
                else:
                    results["failed"] += len(emails)
                    error_msg = result.get("error", f"Status code: {result.get('status_code', 'unknown')}")
                    for email in emails:
                        results["errors"].append({
                            "email": email,
                            "error": error_msg
                        })
            
            # Delay between batches (except for the last batch)
            if i + args.batch_size < total and not args.dry_run:
                time.sleep(args.delay)
    
    elapsed = time.time() - start_time
    
    print()
    print("=" * 60)
    print("📊 Processing Complete!")
    print(f"   Total: {results['total']}")
    print(f"   Successful: {results['successful']}")
    print(f"   Failed: {results['failed']}")
    print(f"   Time elapsed: {elapsed:.2f} seconds")
    
    if results["errors"]:
        print(f"\n⚠️  Errors ({len(results['errors'])}):")
        for error in results["errors"][:10]:  # Show first 10 errors
            print(f"   Row {error['row']}: {error['email']} - {error['error']}")
        if len(results["errors"]) > 10:
            print(f"   ... and {len(results['errors']) - 10} more errors")
    
    if args.dry_run:
        print("\n🔍 This was a dry run. Use without --dry-run to actually unsubscribe.")

if __name__ == "__main__":
    main()

File: subscription-functions-copy/api/bulk-unsubscribe.js
/**
 * Bulk Unsubscribe Endpoint
 * Unsubscribes multiple emails from one or more brands
 * 
 * POST /api/bulk-unsubscribe
 * Body: {
 *   "emails": ["email1@example.com", "email2@example.com"],
 *   "brand": "heebnewsletters" or ["heebnewsletters", "thepicklereport"],
 *   "options": {
 *     "batchSize": 10,
 *     "delayBetweenBatches": 500
 *   }
 * }
 * 
 * Or with per-email brands:
 * {
 *   "unsubscribes": [
 *     { "email": "user1@example.com", "brands": ["heebnewsletters"] },
 *     { "email": "user2@example.com", "brands": ["heebnewsletters", "thepicklereport"] }
 *   ]
 * }
 * 
 * Or CSV format:
 * Body: "email,brand\nuser1@example.com,heebnewsletters\nuser2@example.com,heebnewsletters"
 */

import { BigQuery } from '@google-cloud/bigquery';
import axios from 'axios';
import crypto from 'crypto';

const bq = new BigQuery({
  projectId: process.env.GCP_PROJECT_ID,
  credentials: process.env.GCP_SERVICE_ACCOUNT_KEY 
    ? JSON.parse(process.env.GCP_SERVICE_ACCOUNT_KEY)
    : undefined
});

// Customer.io API configuration
const CIO_SITE_ID = process.env.CIO_SITE_ID;
const CIO_API_KEY = process.env.CIO_API_KEY;
const CIO_TRACK_URL = process.env.CIO_TRACK_URL || 'https://track.customer.io/api/v2';

// Create basic auth header
const cioAuth = Buffer.from(`${CIO_SITE_ID}:${CIO_API_KEY}`).toString('base64');

// Mapping from brand IDs to Customer.io brand object names
const CUSTOMER_IO_BRAND_NAMES = {
  'batmitzvahhorrorstories': 'Bat Mitzvah Horror Stories',
  'grapejuiceandnostalgia': 'Grape Juice and Nostalgia',
  'hardresets': 'Hard Resets',
  'highdiaries': 'High Diaries',
  'hipspeak': 'Hipspeak',
  'hookuplists': 'Hookup Lists',
  'millennialvsgenz': 'Millennial vs Gen Z',
  'obscuremixtape': 'Obscure Mixtape',
  'onetimeatcamp': 'One Time at Camp',
  'the90sparent': 'The 90s Parent',
  'thecomingofageparty': 'The Coming of Age Party',
  'thedadsdad': 'The Dad\'s Dad',
  'theeyeballerscookbook': 'The Eyeballer\'s Cookbook',
  'themixedhome': 'The Mixed Home',
  'thepackandplay': 'The Pack and Play',
  'thepicklereport': 'The Pickle Report',
  'theproudparent': 'The Proud Parent',
  'thequirkiest': 'The Quirkiest',
  'thestewardprize': 'The Steward Prize',
  'toddlercinema': 'Toddler Cinema',
  'zitsandcake': 'Zits and Cake',
  'heebnewsletters': 'Heeb Newsletters'
};

// HTTP client with timeout
const httpClient = axios.create({
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json'
  }
});

async function findUserByEmail(email) {
  const query = `
    SELECT userID 
    FROM \`${process.env.GCP_PROJECT_ID}.analytics.users\`
    WHERE email = @email 
    LIMIT 1
  `;
  
  try {
    const [rows] = await bq.query({ query, params: { email: email.toLowerCase() } });
    return rows.length > 0 ? rows[0].userID : null;
  } catch (error) {
    console.error('User lookup error:', error.message);
    return null;
  }
}

/**
 * Send batched requests to Customer.io
 */
async function sendBatchedCustomerIORequests(batchRequests) {
  try {
    const response = await httpClient.post(`${CIO_TRACK_URL}/batch`, {
      batch: batchRequests
    }, {
      headers: {
        'Authorization': `Basic ${cioAuth}`
      }
    });

    return { success: true, response: response.data };
    
  } catch (error) {
    console.error('❌ Batch request failed:', error.message);
    return { success: false, error: error.message };
  }
}

/**
 * Remove brand relationships from Customer.io
 */
async function removeBrandRelationships(userID, brandIds, normalizedEmail) {
  const relationships = brandIds.map(brandId => ({
    identifiers: {
      object_type_id: "1",
      object_id: CUSTOMER_IO_BRAND_NAMES[brandId] || brandId
    }
  }));
  
  // Prepare unsubscription attributes
  const unsubscribeAttributes = {};
  const unsubscribeDate = new Date().getTime();
  
  // Add unsubscribed_brands attribute
  const unsubscribedBrandsObj = {};
  brandIds.forEach(brandId => {
    unsubscribedBrandsObj[brandId] = unsubscribeDate;
  });
  unsubscribeAttributes['unsubscribed_brands'] = unsubscribedBrandsObj;
  
  // Remove individual brand subscription attributes
  brandIds.forEach(brandId => {
    unsubscribeAttributes[`subscribed_to_${brandId}`] = false;
    unsubscribeAttributes[`${brandId}_subscription_date`] = null;
    unsubscribeAttributes[`lead_for_${brandId}`] = false;
  });
  
  // Create batched requests for unsubscribe
  const attributeRequest = {
    type: "person",
    identifiers: { id: userID },
    action: "identify",
    attributes: {
      email: normalizedEmail,
      ...unsubscribeAttributes
    }
  };
  
  const relationshipRequest = {
    type: "person",
    identifiers: { id: userID },
    action: "delete_relationships",
    cio_relationships: relationships
  };
  
  try {
    const batchRequests = [attributeRequest, relationshipRequest];
    const result = await sendBatchedCustomerIORequests(batchRequests);
    
    if (result.success) {
      return { 
        success: true, 
        removed_brands: brandIds,
        unsubscribe_date: new Date(unsubscribeDate).toISOString()
      };
    } else {
      return { success: false, error: result.error };
    }
  } catch (error) {
    return { success: false, error: error.message };
  }
}

/**
 * Remove brand from subscriptions and log unsubscription to BigQuery
 */
async function removeBrandFromSubscriptionsAndLogUnsubscription(userId, brandId, unsubscribeDate, normalizedEmail) {
  try {
    // First, get current user data to update both subscriptions and unsubscribed_brands
    const query = `
      SELECT userID, subscriptions, unsubscribed_brands
      FROM \`${process.env.GCP_PROJECT_ID}.analytics.users\`
      WHERE userID = @userID 
      LIMIT 1
    `;
    
    const [rows] = await bq.query({ query, params: { userID: userId } });
    
    if (rows.length === 0) {
      return { success: false, error: 'User not found' };
    }
    
    const currentUser = rows[0];
    let subscriptions = {};
    let unsubscribedBrands = {};
    
    // Parse existing subscriptions if they exist
    if (currentUser.subscriptions) {
      try {
        subscriptions = typeof currentUser.subscriptions === 'string' 
          ? JSON.parse(currentUser.subscriptions) 
          : currentUser.subscriptions;
      } catch (e) {
        subscriptions = {};
      }
    }
    
    // Parse existing unsubscribed_brands if they exist
    if (currentUser.unsubscribed_brands) {
      try {
        unsubscribedBrands = typeof currentUser.unsubscribed_brands === 'string'
          ? JSON.parse(currentUser.unsubscribed_brands)
          : currentUser.unsubscribed_brands;
      } catch (e) {
        unsubscribedBrands = {};
      }
    }
    
    // Remove brand from subscriptions
    if (subscriptions[brandId]) {
      delete subscriptions[brandId];
    }
    
    // Add/update unsubscription entry for this brand
    unsubscribedBrands[brandId] = new Date(unsubscribeDate).getTime();
    
    // Use MERGE to update
    const mergeQuery = `
      MERGE \`${process.env.GCP_PROJECT_ID}.analytics.users\` AS target
      USING (
        SELECT 
          @userID as userID,
          PARSE_JSON(@subscriptions) as subscriptions,
          @unsubscribed_brands as unsubscribed_brands
      ) AS source
      ON target.userID = source.userID
      WHEN MATCHED THEN
        UPDATE SET 
          subscriptions = source.subscriptions,
          unsubscribed_brands = source.unsubscribed_brands
    `;
    
    await bq.query({
      query: mergeQuery,
      params: {
        userID: userId,
        subscriptions: JSON.stringify(subscriptions),
        unsubscribed_brands: JSON.stringify(unsubscribedBrands)
      },
      types: {
        userID: 'STRING',
        subscriptions: 'STRING',
        unsubscribed_brands: 'STRING'
      }
    });
    
    return { success: true };
    
  } catch (error) {
    console.error('❌ BigQuery subscription/unsubscription update error:', error.message);
    return { success: false, error: error.message };
  }
}

/**
 * Process a single unsubscribe
 */
async function processSingleUnsubscribe(email, brandIds) {
  const normalizedEmail = email.toLowerCase().trim();
  const brandList = Array.isArray(brandIds) ? brandIds : [brandIds].filter(Boolean);
  
  if (brandList.length === 0) {
    return {
      email: normalizedEmail,
      success: false,
      error: 'No brands specified'
    };
  }
  
  try {
    // Find user by email
    const userID = await findUserByEmail(normalizedEmail);
    
    if (!userID) {
      return {
        email: normalizedEmail,
        success: false,
        error: 'User not found'
      };
    }
    
    // Process unsubscription for each brand
    const results = [];
    for (const brandId of brandList) {
      // Remove brand relationships from Customer.io
      const cioResult = await removeBrandRelationships(userID, [brandId], normalizedEmail);
      
      if (cioResult.success) {
        // Remove brand from BigQuery subscriptions AND log unsubscription
        const bqResult = await removeBrandFromSubscriptionsAndLogUnsubscription(
          userID, 
          brandId, 
          cioResult.unsubscribe_date,
          normalizedEmail
        );
        
        if (bqResult.success) {
          results.push({ brand: brandId, success: true });
        } else {
          results.push({ brand: brandId, success: false, error: `BigQuery update failed: ${bqResult.error}` });
        }
      } else {
        results.push({ brand: brandId, success: false, error: `Customer.io update failed: ${cioResult.error}` });
      }
    }
    
    const allSuccessful = results.every(r => r.success);
    
    return {
      email: normalizedEmail,
      success: allSuccessful,
      brands: results,
      error: allSuccessful ? null : 'Some brands failed to unsubscribe'
    };
    
  } catch (error) {
    return {
      email: normalizedEmail,
      success: false,
      error: error.message || 'Unknown error'
    };
  }
}

function parseCSV(csvText) {
  const lines = csvText.split('\n').filter(line => line.trim());
  if (lines.length < 2) {
    throw new Error('CSV must have at least a header and one data row');
  }
  
  const headerLine = lines[0];
  const header = headerLine.split(',').map(h => h.toLowerCase().trim());
  
  const emailIndex = header.indexOf('email');
  if (emailIndex === -1) {
    throw new Error('CSV must have an "email" column');
  }
  
  const brandIndex = header.indexOf('brand');
  if (brandIndex === -1) {
    throw new Error('CSV must have a "brand" column');
  }
  
  const unsubscribes = [];
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;
    
    const parts = line.split(',').map(p => p.trim());
    const email = parts[emailIndex];
    const brand = parts[brandIndex];
    
    if (!email || !brand) continue;
    
    unsubscribes.push({
      email: email,
      brands: [brand.toLowerCase().trim()]
    });
  }
  
  return unsubscribes;
}

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed. Use POST.' });
  }
  
  let unsubscribeList = [];
  
  // Check if body is CSV text or JSON
  if (typeof req.body === 'string' && req.body.includes(',')) {
    // Assume it's CSV text
    console.log('📄 Parsing CSV from request body...');
    try {
      unsubscribeList = parseCSV(req.body);
    } catch (error) {
      return res.status(400).json({ error: `CSV parsing error: ${error.message}` });
    }
  } else if (req.body.unsubscribes && Array.isArray(req.body.unsubscribes)) {
    // Advanced format: per-email brands
    unsubscribeList = req.body.unsubscribes.map(item => ({
      email: item.email,
      brands: Array.isArray(item.brands) ? item.brands : [item.brands].filter(Boolean)
    }));
  } else if (req.body.emails && req.body.brand) {
    // Simple format: same brand(s) for all emails
    const brandList = Array.isArray(req.body.brand) ? req.body.brand : [req.body.brand].filter(Boolean);
    if (brandList.length === 0) {
      return res.status(400).json({ error: 'No brands specified' });
    }
    
    unsubscribeList = req.body.emails.map(email => ({
      email: email,
      brands: brandList
    }));
  } else {
    return res.status(400).json({
      error: 'Invalid request body',
      usage: {
        simple: {
          emails: ['email1@example.com', 'email2@example.com'],
          brand: 'heebnewsletters' // or ['heebnewsletters', 'thepicklereport']
        },
        advanced: {
          unsubscribes: [
            { email: 'user1@example.com', brands: ['heebnewsletters'] },
            { email: 'user2@example.com', brands: ['heebnewsletters', 'thepicklereport'] }
          ]
        },
        csv: 'email,brand\nuser1@example.com,heebnewsletters\nuser2@example.com,heebnewsletters'
      }
    });
  }
  
  if (unsubscribeList.length === 0) {
    return res.status(400).json({ error: 'No unsubscribes to process' });
  }
  
  const options = req.body.options || {};
  const batchSize = options.batchSize || 10;
  const delayBetweenBatches = options.delayBetweenBatches || 500;
  const dryRun = options.dryRun || false;
  
  console.log(`🚀 Starting bulk unsubscribe processing: ${unsubscribeList.length} emails`);
  console.log(`📦 Batch size: ${batchSize}, Delay: ${delayBetweenBatches}ms`);
  
  if (dryRun) {
    return res.status(200).json({
      success: true,
      dryRun: true,
      message: `Would process ${unsubscribeList.length} unsubscribes`,
      unsubscribes: unsubscribeList
    });
  }
  
  const results = {
    total: unsubscribeList.length,
    successful: 0,
    failed: 0,
    errors: []
  };
  
  // Process in batches
  for (let i = 0; i < unsubscribeList.length; i += batchSize) {
    const batch = unsubscribeList.slice(i, i + batchSize);
    const batchNum = Math.floor(i / batchSize) + 1;
    const totalBatches = Math.ceil(unsubscribeList.length / batchSize);
    
    console.log(`\n📦 Processing batch ${batchNum}/${totalBatches} (${batch.length} emails)`);
    
    // Process batch in parallel
    const batchPromises = batch.map(item => 
      processSingleUnsubscribe(item.email, item.brands)
    );
    
    const batchResults = await Promise.all(batchPromises);
    
    for (const result of batchResults) {
      if (result.success) {
        results.successful++;
        console.log(`✅ ${result.email}: Unsubscribed from ${result.brands.length} brand(s)`);
      } else {
        results.failed++;
        results.errors.push({
          email: result.email,
          error: result.error
        });
        console.log(`❌ ${result.email}: ${result.error}`);
      }
    }
    
    // Delay between batches (except for the last batch)
    if (i + batchSize < unsubscribeList.length) {
      await new Promise(resolve => setTimeout(resolve, delayBetweenBatches));
    }
  }
  
  console.log(`\n📊 Processing Complete!`);
  console.log(`Total: ${results.total}, Successful: ${results.successful}, Failed: ${results.failed}`);
  
  return res.status(200).json({
    success: true,
    summary: results,
    message: `Processed ${results.total} unsubscribes: ${results.successful} successful, ${results.failed} failed`
  });
}
